File: services/collector/src/resource-audit-config.ts
import fs from "node:fs";
import path from "node:path";

export const RESOURCE_AUDIT_HINT_TYPES = [
  "pii_hint",
  "secret_hint",
  "credential_hint",
  "config_hint",
  "database_hint",
] as const;

export type ResourceAuditHintType = (typeof RESOURCE_AUDIT_HINT_TYPES)[number];

export type ResourceAuditConfig = {
  dangerousPathRules: {
    posixPrefixes: string[];
    windowsPrefixes: string[];
    windowsRegex: string[];
    caseInsensitive: boolean;
  };
  largeRead: {
    thresholdChars: number;
  };
  largeToolResult: {
    thresholdChars: number;
  };
  policyLink: {
    enabled: boolean;
    targetActions: string[];
    matchScope: "span" | "trace";
  };
  policyHintTypes: {
    enabledHintTypes: ResourceAuditHintType[];
    defaultHintType: ResourceAuditHintType | null;
    includeUnlabeledPolicyHit: boolean;
  };
  shellExec: {
    loopAlerts: {
      minRepeatCount: number;
      maxItems: number;
    };
    tokenRisks: {
      stdoutCharsThreshold: number;
      maxItems: number;
    };
    commandSemantics: {
      enabled: boolean;
      defaultPlatform: "unix" | "windows_cmd" | "powershell";
      platformDetect: {
        preferSpanNameHints: boolean;
        spanNameHints: {
          unix: string[];
          windows_cmd: string[];
          powershell: string[];
        };
      };
      aliases: Record<string, string>;
      categories: {
        file: string[];
        network: string[];
        system: string[];
        process: string[];
        package: string[];
      };
      readLikeCommands: string[];
      diagnosticPatterns: {
        commandNotFound: string[];
        permissionDenied: string[];
        illegalArgHint: string[];
      };
    };
  };
};

const DEFAULT_CONFIG: ResourceAuditConfig = {
  dangerousPathRules: {
    posixPrefixes: ["/etc", "/root", "/var/lib", "/home", "~/.ssh", ".env", "private.key"],
    windowsPrefixes: [
      "c:\\users\\",
      "c:\\windows\\",
      "c:\\programdata\\",
      "%appdata%\\",
      "%userprofile%\\",
      "\\\\",
    ],
    windowsRegex: [
      "^[a-zA-Z]:\\\\\\\\Users\\\\\\\\[^\\\\\\\\]+\\\\\\\\\\.ssh\\\\\\\\",
      "\\\\\\\\[^\\\\\\\\]+\\\\\\\\[^\\\\\\\\]+",
      "^%APPDATA%\\\\\\\\",
      "^%USERPROFILE%\\\\\\\\",
    ],
    caseInsensitive: true,
  },
  largeRead: {
    thresholdChars: 500_000,
  },
  largeToolResult: {
    thresholdChars: 8_192,
  },
  policyLink: {
    enabled: true,
    targetActions: [],
    matchScope: "span",
  },
  policyHintTypes: {
    enabledHintTypes: [...RESOURCE_AUDIT_HINT_TYPES],
    defaultHintType: null,
    includeUnlabeledPolicyHit: false,
  },
  shellExec: {
    loopAlerts: {
      minRepeatCount: 3,
      maxItems: 20,
    },
    tokenRisks: {
      stdoutCharsThreshold: 24_000,
      maxItems: 15,
    },
    commandSemantics: {
      enabled: true,
      defaultPlatform: "unix",
      platformDetect: {
        preferSpanNameHints: true,
        spanNameHints: {
          unix: ["bash", "zsh", "sh", "terminal", "shell"],
          windows_cmd: ["cmd", "cmd.exe", "run_cmd", "runcmd"],
          powershell: ["pwsh", "powershell"],
        },
      },
      aliases: {
        ls: "get-childitem",
        dir: "get-childitem",
        gci: "get-childitem",
        cat: "get-content",
        type: "get-content",
        gc: "get-content",
        rm: "remove-item",
        del: "remove-item",
        erase: "remove-item",
        cp: "copy-item",
        copy: "copy-item",
        mv: "move-item",
        move: "move-item",
      },
      categories: {
        file: [
          "ls", "cat", "head", "tail", "less", "more", "find", "grep", "rg", "fd", "cp", "mv", "rm", "mkdir", "rmdir",
          "touch", "chmod", "chown", "stat", "diff", "wc", "sort", "uniq", "tee", "xargs", "sed", "awk", "readlink",
          "realpath", "tree", "dir", "type", "copy", "move", "del", "erase", "findstr", "where",
          "get-childitem", "get-content", "set-content", "copy-item", "move-item", "remove-item", "select-string",
        ],
        network: ["curl", "wget", "ping", "ssh", "scp", "rsync", "nc", "netcat", "telnet", "dig", "nslookup", "invoke-webrequest", "irm"],
        system: ["sudo", "su", "systemctl", "service", "mount", "umount", "df", "du", "free", "uname", "whoami", "id", "env", "printenv", "export", "ulimit", "sysctl", "setx"],
        process: ["ps", "top", "htop", "kill", "killall", "pkill", "pgrep", "jobs", "fg", "bg", "nohup", "nice", "tasklist", "taskkill", "get-process", "stop-process"],
        package: ["npm", "pnpm", "yarn", "bun", "pip", "pip3", "apt", "apt-get", "yum", "dnf", "brew", "cargo", "go", "choco", "winget"],
      },
      readLikeCommands: ["cat", "head", "tail", "less", "grep", "rg", "find", "type", "findstr", "get-content", "select-string"],
      diagnosticPatterns: {
        commandNotFound: ["command not found", "not found as command", "is not recognized as an internal or external command"],
        permissionDenied: ["permission denied", "access is denied", "operation not permitted", "eacces"],
        illegalArgHint: ["illegal option", "invalid option", "unrecognized option", "syntax error", "usage:", "parameter cannot be found"],
      },
    },
  },
};

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function cleanStringArray(v: unknown): string[] {
  if (!Array.isArray(v)) {
    return [];
  }
  return v.map((x) => String(x ?? "").trim()).filter((x) => x.length > 0);
}

function cleanHintTypes(v: unknown): ResourceAuditHintType[] {
  const set = new Set<ResourceAuditHintType>();
  for (const x of cleanStringArray(v)) {
    if ((RESOURCE_AUDIT_HINT_TYPES as readonly string[]).includes(x)) {
      set.add(x as ResourceAuditHintType);
    }
  }
  return [...set];
}

function cleanStringMap(v: unknown): Record<string, string> {
  if (!isRecord(v)) {
    return {};
  }
  const out: Record<string, string> = {};
  for (const [k, val] of Object.entries(v)) {
    const key = String(k ?? "").trim().toLowerCase();
    const next = String(val ?? "").trim().toLowerCase();
    if (!key || !next) {
      continue;
    }
    out[key] = next;
  }
  return out;
}

export function defaultResourceAuditConfig(): ResourceAuditConfig {
  return JSON.parse(JSON.stringify(DEFAULT_CONFIG)) as ResourceAuditConfig;
}

export function normalizeResourceAuditConfig(raw: unknown): ResourceAuditConfig {
  const base = defaultResourceAuditConfig();
  if (!isRecord(raw)) {
    return base;
  }
  const dangerous = isRecord(raw.dangerousPathRules) ? raw.dangerousPathRules : {};
  const largeRead = isRecord(raw.largeRead) ? raw.largeRead : {};
  const largeToolResult = isRecord(raw.largeToolResult) ? raw.largeToolResult : {};
  const policyLink = isRecord(raw.policyLink) ? raw.policyLink : {};
  const policyHintTypes = isRecord(raw.policyHintTypes) ? raw.policyHintTypes : {};
  const shellExec = isRecord(raw.shellExec) ? raw.shellExec : {};
  const shellLoopAlerts = isRecord(shellExec.loopAlerts) ? shellExec.loopAlerts : {};
  const shellTokenRisks = isRecord(shellExec.tokenRisks) ? shellExec.tokenRisks : {};
  const commandSemantics = isRecord(shellExec.commandSemantics) ? shellExec.commandSemantics : {};
  const platformDetect = isRecord(commandSemantics.platformDetect) ? commandSemantics.platformDetect : {};
  const spanNameHints = isRecord(platformDetect.spanNameHints) ? platformDetect.spanNameHints : {};
  const categories = isRecord(commandSemantics.categories) ? commandSemantics.categories : {};
  const diagnosticPatterns = isRecord(commandSemantics.diagnosticPatterns) ? commandSemantics.diagnosticPatterns : {};

  const next: ResourceAuditConfig = {
    dangerousPathRules: {
      posixPrefixes: cleanStringArray(dangerous.posixPrefixes),
      windowsPrefixes: cleanStringArray(dangerous.windowsPrefixes),
      windowsRegex: cleanStringArray(dangerous.windowsRegex),
      caseInsensitive:
        typeof dangerous.caseInsensitive === "boolean"
          ? dangerous.caseInsensitive
          : base.dangerousPathRules.caseInsensitive,
    },
    largeRead: {
      thresholdChars:
        Number.isFinite(Number(largeRead.thresholdChars)) && Number(largeRead.thresholdChars) >= 0
          ? Math.floor(Number(largeRead.thresholdChars))
          : base.largeRead.thresholdChars,
    },
    largeToolResult: {
      thresholdChars:
        Number.isFinite(Number(largeToolResult.thresholdChars)) && Number(largeToolResult.thresholdChars) >= 0
          ? Math.floor(Number(largeToolResult.thresholdChars))
          : base.largeToolResult.thresholdChars,
    },
    policyLink: {
      enabled: typeof policyLink.enabled === "boolean" ? policyLink.enabled : base.policyLink.enabled,
      targetActions: cleanStringArray(policyLink.targetActions),
      matchScope: policyLink.matchScope === "trace" ? "trace" : "span",
    },
    policyHintTypes: {
      enabledHintTypes: cleanHintTypes(policyHintTypes.enabledHintTypes),
      defaultHintType:
        typeof policyHintTypes.defaultHintType === "string" &&
        (RESOURCE_AUDIT_HINT_TYPES as readonly string[]).includes(policyHintTypes.defaultHintType)
          ? (policyHintTypes.defaultHintType as ResourceAuditHintType)
          : null,
      includeUnlabeledPolicyHit:
        typeof policyHintTypes.includeUnlabeledPolicyHit === "boolean"
          ? policyHintTypes.includeUnlabeledPolicyHit
          : base.policyHintTypes.includeUnlabeledPolicyHit,
    },
    shellExec: {
      loopAlerts: {
        minRepeatCount:
          Number.isFinite(Number(shellLoopAlerts.minRepeatCount)) && Number(shellLoopAlerts.minRepeatCount) >= 1
            ? Math.floor(Number(shellLoopAlerts.minRepeatCount))
            : base.shellExec.loopAlerts.minRepeatCount,
        maxItems:
          Number.isFinite(Number(shellLoopAlerts.maxItems)) && Number(shellLoopAlerts.maxItems) >= 1
            ? Math.floor(Number(shellLoopAlerts.maxItems))
            : base.shellExec.loopAlerts.maxItems,
      },
      tokenRisks: {
        stdoutCharsThreshold:
          Number.isFinite(Number(shellTokenRisks.stdoutCharsThreshold)) &&
          Number(shellTokenRisks.stdoutCharsThreshold) >= 0
            ? Math.floor(Number(shellTokenRisks.stdoutCharsThreshold))
            : base.shellExec.tokenRisks.stdoutCharsThreshold,
        maxItems:
          Number.isFinite(Number(shellTokenRisks.maxItems)) && Number(shellTokenRisks.maxItems) >= 1
            ? Math.floor(Number(shellTokenRisks.maxItems))
            : base.shellExec.tokenRisks.maxItems,
      },
      commandSemantics: {
        enabled:
          typeof commandSemantics.enabled === "boolean"
            ? commandSemantics.enabled
            : base.shellExec.commandSemantics.enabled,
        defaultPlatform:
          commandSemantics.defaultPlatform === "windows_cmd" || commandSemantics.defaultPlatform === "powershell"
            ? commandSemantics.defaultPlatform
            : "unix",
        platformDetect: {
          preferSpanNameHints:
            typeof platformDetect.preferSpanNameHints === "boolean"
              ? platformDetect.preferSpanNameHints
              : base.shellExec.commandSemantics.platformDetect.preferSpanNameHints,
          spanNameHints: {
            unix: cleanStringArray(spanNameHints.unix),
            windows_cmd: cleanStringArray(spanNameHints.windows_cmd),
            powershell: cleanStringArray(spanNameHints.powershell),
          },
        },
        aliases: cleanStringMap(commandSemantics.aliases),
        categories: {
          file: cleanStringArray(categories.file),
          network: cleanStringArray(categories.network),
          system: cleanStringArray(categories.system),
          process: cleanStringArray(categories.process),
          package: cleanStringArray(categories.package),
        },
        readLikeCommands: cleanStringArray(commandSemantics.readLikeCommands),
        diagnosticPatterns: {
          commandNotFound: cleanStringArray(diagnosticPatterns.commandNotFound),
          permissionDenied: cleanStringArray(diagnosticPatterns.permissionDenied),
          illegalArgHint: cleanStringArray(diagnosticPatterns.illegalArgHint),
        },
      },
    },
  };
  if (Object.keys(next.shellExec.commandSemantics.aliases).length === 0) {
    next.shellExec.commandSemantics.aliases = { ...base.shellExec.commandSemantics.aliases };
  }
  if (next.shellExec.commandSemantics.categories.file.length === 0) {
    next.shellExec.commandSemantics.categories = JSON.parse(
      JSON.stringify(base.shellExec.commandSemantics.categories),
    ) as ResourceAuditConfig["shellExec"]["commandSemantics"]["categories"];
  }
  if (next.shellExec.commandSemantics.readLikeCommands.length === 0) {
    next.shellExec.commandSemantics.readLikeCommands = [
      ...base.shellExec.commandSemantics.readLikeCommands,
    ];
  }
  if (next.policyHintTypes.enabledHintTypes.length === 0) {
    next.policyHintTypes.enabledHintTypes = [...base.policyHintTypes.enabledHintTypes];
  }
  for (const pat of next.dangerousPathRules.windowsRegex) {
    // Validate regex eagerly to reject bad config writes.
    new RegExp(pat);
  }
  return next;
}

function resolveResourceAuditConfigPath(): string {
  const raw = process.env.CRABAGENT_RESOURCE_AUDIT_CONFIG?.trim();
  if (raw) {
    return path.resolve(raw);
  }
  return path.resolve(process.cwd(), "resource-audit.config.json");
}

let cachedPath = "";
let cachedMtimeMs = -1;
let cachedConfig: ResourceAuditConfig | null = null;

export function loadResourceAuditConfig(): ResourceAuditConfig {
  const configPath = resolveResourceAuditConfigPath();
  try {
    const st = fs.statSync(configPath);
    const nextMtimeMs = Number(st.mtimeMs) || 0;
    if (cachedConfig && cachedPath === configPath && nextMtimeMs === cachedMtimeMs) {
      return JSON.parse(JSON.stringify(cachedConfig)) as ResourceAuditConfig;
    }
    const text = fs.readFileSync(configPath, "utf8");
    const normalized = normalizeResourceAuditConfig(JSON.parse(text));
    cachedPath = configPath;
    cachedMtimeMs = nextMtimeMs;
    cachedConfig = normalized;
    return JSON.parse(JSON.stringify(normalized)) as ResourceAuditConfig;
  } catch {
    return defaultResourceAuditConfig();
  }
}

File: iseeagentc/model/resource_audit_config.go
package model

import (
	"encoding/json"
	"errors"
	"math"
	"regexp"
	"strings"

	"iseeagentc/conf"

	toml "github.com/pelletier/go-toml/v2"
)

// ResourceAuditQueryConfig 与 Node ResourceAuditConfig 对齐（JSON camelCase）；资源审计查询使用该配置。
type ResourceAuditQueryConfig struct {
	DangerousPathRules struct {
		PosixPrefixes   []string `json:"posixPrefixes"`
		WindowsPrefixes []string `json:"windowsPrefixes"`
		WindowsRegex    []string `json:"windowsRegex"`
		CaseInsensitive bool     `json:"caseInsensitive"`
	} `json:"dangerousPathRules"`
	LargeRead struct {
		ThresholdChars int `json:"thresholdChars"`
	} `json:"largeRead"`
	LargeToolResult struct {
		ThresholdChars int `json:"thresholdChars"`
	} `json:"largeToolResult"`
	PolicyLink struct {
		Enabled       bool     `json:"enabled"`
		TargetActions []string `json:"targetActions"`
		MatchScope    string   `json:"matchScope"` // "span" | "trace"
	} `json:"policyLink"`
	ShellExec struct {
		LoopAlerts struct {
			MinRepeatCount int `json:"minRepeatCount"`
			MaxItems       int `json:"maxItems"`
		} `json:"loopAlerts"`
		TokenRisks struct {
			StdoutCharsThreshold int `json:"stdoutCharsThreshold"`
			MaxItems             int `json:"maxItems"`
		} `json:"tokenRisks"`
		CommandSemantics struct {
			Enabled         bool   `json:"enabled"`
			DefaultPlatform string `json:"defaultPlatform"`
			PlatformDetect  struct {
				PreferSpanNameHints bool `json:"preferSpanNameHints"`
				SpanNameHints       struct {
					Unix       []string `json:"unix"`
					WindowsCmd []string `json:"windows_cmd"`
					Powershell []string `json:"powershell"`
				} `json:"spanNameHints"`
			} `json:"platformDetect"`
			Aliases    map[string]string `json:"aliases"`
			Categories struct {
				File    []string `json:"file"`
				Network []string `json:"network"`
				System  []string `json:"system"`
				Process []string `json:"process"`
				Package []string `json:"package"`
			} `json:"categories"`
			ReadLikeCommands   []string `json:"readLikeCommands"`
			DiagnosticPatterns struct {
				CommandNotFound  []string `json:"commandNotFound"`
				PermissionDenied []string `json:"permissionDenied"`
				IllegalArgHint   []string `json:"illegalArgHint"`
			} `json:"diagnosticPatterns"`
		} `json:"commandSemantics"`
	} `json:"shellExec"`
}

func loadBaseFromEmbedded() (ResourceAuditQueryConfig, error) {
	b, err := decodeConfigBytesToJSON(conf.DefaultResourceAuditConfigTOML())
	if err != nil {
		return ResourceAuditQueryConfig{}, err
	}
	var c ResourceAuditQueryConfig
	if err := json.Unmarshal(b, &c); err != nil {
		return ResourceAuditQueryConfig{}, err
	}
	return c, nil
}

func decodeConfigBytesToJSON(raw []byte) ([]byte, error) {
	var anyMap map[string]any
	if err := toml.Unmarshal(raw, &anyMap); err == nil {
		return json.Marshal(anyMap)
	}
	if err := json.Unmarshal(raw, &anyMap); err == nil {
		return json.Marshal(anyMap)
	}
	return nil, errors.New("invalid config format: expect TOML or JSON object")
}

// DefaultResourceAuditQueryConfig 等价于 Node defaultResourceAuditConfig()（内嵌 JSON 的副本）。
func DefaultResourceAuditQueryConfig() ResourceAuditQueryConfig {
	c, err := loadBaseFromEmbedded()
	if err != nil {
		return ResourceAuditQueryConfig{}
	}
	return c
}

func getRawObject(m map[string]json.RawMessage, key string) map[string]json.RawMessage {
	v, ok := m[key]
	if !ok || len(v) == 0 {
		return map[string]json.RawMessage{}
	}
	var obj map[string]json.RawMessage
	if json.Unmarshal(v, &obj) != nil {
		return map[string]json.RawMessage{}
	}
	return obj
}

func cleanStringArrayFromRaw(raw json.RawMessage) []string {
	if len(raw) == 0 {
		return nil
	}
	var arr []any
	if err := json.Unmarshal(raw, &arr); err != nil {
		return nil
	}
	var out []string
	for _, x := range arr {
		s := strings.TrimSpace(stringFromJSONScalar(x))
		if s != "" {
			out = append(out, s)
		}
	}
	return out
}

func stringFromJSONScalar(x any) string {
	switch v := x.(type) {
	case string:
		return v
	default:
		b, _ := json.Marshal(v)
		return strings.TrimSpace(string(b))
	}
}

func cleanStringMapFromRaw(raw json.RawMessage) map[string]string {
	if len(raw) == 0 {
		return map[string]string{}
	}
	var m map[string]any
	if json.Unmarshal(raw, &m) != nil {
		return map[string]string{}
	}
	out := make(map[string]string)
	for k, val := range m {
		key := strings.ToLower(strings.TrimSpace(k))
		next := strings.ToLower(strings.TrimSpace(stringFromJSONScalar(val)))
		if key == "" || next == "" {
			continue
		}
		out[key] = next
	}
	return out
}

func intFromRawGEZero(raw json.RawMessage, fallback int) int {
	if len(raw) == 0 {
		return fallback
	}
	var n float64
	if err := json.Unmarshal(raw, &n); err != nil {
		return fallback
	}
	if !isFiniteFloatForConfig(n) || n < 0 {
		return fallback
	}
	return int(n)
}

func intFromRawGEFloor1(raw json.RawMessage, fallback int) int {
	if len(raw) == 0 {
		return fallback
	}
	var n float64
	if err := json.Unmarshal(raw, &n); err != nil {
		return fallback
	}
	if !isFiniteFloatForConfig(n) || n < 1 {
		return fallback
	}
	return int(n)
}

func isFiniteFloatForConfig(f float64) bool {
	return !math.IsNaN(f) && !math.IsInf(f, 0)
}

func boolFromRaw(raw json.RawMessage, fallback bool) bool {
	if len(raw) == 0 {
		return fallback
	}
	var b bool
	if err := json.Unmarshal(raw, &b); err != nil {
		return fallback
	}
	return b
}

func stringFromRaw(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		return ""
	}
	return strings.TrimSpace(s)
}

func copyCategories(dst *ResourceAuditQueryConfig, src ResourceAuditQueryConfig) {
	dst.ShellExec.CommandSemantics.Categories.File = append([]string(nil), src.ShellExec.CommandSemantics.Categories.File...)
	dst.ShellExec.CommandSemantics.Categories.Network = append([]string(nil), src.ShellExec.CommandSemantics.Categories.Network...)
	dst.ShellExec.CommandSemantics.Categories.System = append([]string(nil), src.ShellExec.CommandSemantics.Categories.System...)
	dst.ShellExec.CommandSemantics.Categories.Process = append([]string(nil), src.ShellExec.CommandSemantics.Categories.Process...)
	dst.ShellExec.CommandSemantics.Categories.Package = append([]string(nil), src.ShellExec.CommandSemantics.Categories.Package...)
}

func copyAliases(dst *ResourceAuditQueryConfig, src ResourceAuditQueryConfig) {
	if src.ShellExec.CommandSemantics.Aliases == nil {
		dst.ShellExec.CommandSemantics.Aliases = map[string]string{}
		return
	}
	m := make(map[string]string, len(src.ShellExec.CommandSemantics.Aliases))
	for k, v := range src.ShellExec.CommandSemantics.Aliases {
		m[k] = v
	}
	dst.ShellExec.CommandSemantics.Aliases = m
}

// normalizeResourceAuditConfigBytes 对齐 Node normalizeResourceAuditConfig；ok 为 false 表示校验失败（如非法正则），调用方应退回默认配置。
func normalizeResourceAuditConfigBytes(b []byte) (ResourceAuditQueryConfig, bool) {
	base, err := loadBaseFromEmbedded()
	if err != nil {
		return ResourceAuditQueryConfig{}, false
	}

	jsonBytes, err := decodeConfigBytesToJSON(b)
	if err != nil {
		return base, true
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(jsonBytes, &raw); err != nil {
		return base, true
	}

	dangerous := getRawObject(raw, "dangerousPathRules")
	largeRead := getRawObject(raw, "largeRead")
	largeToolResult := getRawObject(raw, "largeToolResult")
	policyLink := getRawObject(raw, "policyLink")
	shellExec := getRawObject(raw, "shellExec")
	shellLoopAlerts := getRawObject(shellExec, "loopAlerts")
	shellTokenRisks := getRawObject(shellExec, "tokenRisks")
	commandSemantics := getRawObject(shellExec, "commandSemantics")
	platformDetect := getRawObject(commandSemantics, "platformDetect")
	spanNameHints := getRawObject(platformDetect, "spanNameHints")
	categories := getRawObject(commandSemantics, "categories")
	diagnosticPatterns := getRawObject(commandSemantics, "diagnosticPatterns")

	var next ResourceAuditQueryConfig

	next.DangerousPathRules.PosixPrefixes = cleanStringArrayFromRaw(dangerous["posixPrefixes"])
	next.DangerousPathRules.WindowsPrefixes = cleanStringArrayFromRaw(dangerous["windowsPrefixes"])
	next.DangerousPathRules.WindowsRegex = cleanStringArrayFromRaw(dangerous["windowsRegex"])
	if _, ok := dangerous["caseInsensitive"]; ok {
		next.DangerousPathRules.CaseInsensitive = boolFromRaw(dangerous["caseInsensitive"], base.DangerousPathRules.CaseInsensitive)
	} else {
		next.DangerousPathRules.CaseInsensitive = base.DangerousPathRules.CaseInsensitive
	}

	next.LargeRead.ThresholdChars = intFromRawGEZero(largeRead["thresholdChars"], base.LargeRead.ThresholdChars)
	next.LargeToolResult.ThresholdChars = intFromRawGEZero(largeToolResult["thresholdChars"], base.LargeToolResult.ThresholdChars)

	next.PolicyLink.Enabled = boolFromRaw(policyLink["enabled"], base.PolicyLink.Enabled)
	next.PolicyLink.TargetActions = cleanStringArrayFromRaw(policyLink["targetActions"])
	ms := stringFromRaw(policyLink["matchScope"])
	if ms == "trace" {
		next.PolicyLink.MatchScope = "trace"
	} else {
		next.PolicyLink.MatchScope = "span"
	}

	next.ShellExec.LoopAlerts.MinRepeatCount = intFromRawGEFloor1(shellLoopAlerts["minRepeatCount"], base.ShellExec.LoopAlerts.MinRepeatCount)
	next.ShellExec.LoopAlerts.MaxItems = intFromRawGEFloor1(shellLoopAlerts["maxItems"], base.ShellExec.LoopAlerts.MaxItems)
	next.ShellExec.TokenRisks.StdoutCharsThreshold = intFromRawGEZero(shellTokenRisks["stdoutCharsThreshold"], base.ShellExec.TokenRisks.StdoutCharsThreshold)
	next.ShellExec.TokenRisks.MaxItems = intFromRawGEFloor1(shellTokenRisks["maxItems"], base.ShellExec.TokenRisks.MaxItems)

	next.ShellExec.CommandSemantics.Enabled = boolFromRaw(commandSemantics["enabled"], base.ShellExec.CommandSemantics.Enabled)
	dp := stringFromRaw(commandSemantics["defaultPlatform"])
	if dp == "windows_cmd" || dp == "powershell" {
		next.ShellExec.CommandSemantics.DefaultPlatform = dp
	} else {
		next.ShellExec.CommandSemantics.DefaultPlatform = "unix"
	}
	if _, ok := platformDetect["preferSpanNameHints"]; ok {
		next.ShellExec.CommandSemantics.PlatformDetect.PreferSpanNameHints = boolFromRaw(
			platformDetect["preferSpanNameHints"],
			base.ShellExec.CommandSemantics.PlatformDetect.PreferSpanNameHints,
		)
	} else {
		next.ShellExec.CommandSemantics.PlatformDetect.PreferSpanNameHints = base.ShellExec.CommandSemantics.PlatformDetect.PreferSpanNameHints
	}
	next.ShellExec.CommandSemantics.PlatformDetect.SpanNameHints.Unix = cleanStringArrayFromRaw(spanNameHints["unix"])
	next.ShellExec.CommandSemantics.PlatformDetect.SpanNameHints.WindowsCmd = cleanStringArrayFromRaw(spanNameHints["windows_cmd"])
	next.ShellExec.CommandSemantics.PlatformDetect.SpanNameHints.Powershell = cleanStringArrayFromRaw(spanNameHints["powershell"])

	next.ShellExec.CommandSemantics.Aliases = cleanStringMapFromRaw(commandSemantics["aliases"])
	next.ShellExec.CommandSemantics.Categories.File = cleanStringArrayFromRaw(categories["file"])
	next.ShellExec.CommandSemantics.Categories.Network = cleanStringArrayFromRaw(categories["network"])
	next.ShellExec.CommandSemantics.Categories.System = cleanStringArrayFromRaw(categories["system"])
	next.ShellExec.CommandSemantics.Categories.Process = cleanStringArrayFromRaw(categories["process"])
	next.ShellExec.CommandSemantics.Categories.Package = cleanStringArrayFromRaw(categories["package"])
	next.ShellExec.CommandSemantics.ReadLikeCommands = cleanStringArrayFromRaw(commandSemantics["readLikeCommands"])
	next.ShellExec.CommandSemantics.DiagnosticPatterns.CommandNotFound = cleanStringArrayFromRaw(diagnosticPatterns["commandNotFound"])
	next.ShellExec.CommandSemantics.DiagnosticPatterns.PermissionDenied = cleanStringArrayFromRaw(diagnosticPatterns["permissionDenied"])
	next.ShellExec.CommandSemantics.DiagnosticPatterns.IllegalArgHint = cleanStringArrayFromRaw(diagnosticPatterns["illegalArgHint"])

	if len(next.ShellExec.CommandSemantics.Aliases) == 0 {
		copyAliases(&next, base)
	}
	if len(next.ShellExec.CommandSemantics.Categories.File) == 0 {
		copyCategories(&next, base)
	}
	if len(next.ShellExec.CommandSemantics.ReadLikeCommands) == 0 {
		next.ShellExec.CommandSemantics.ReadLikeCommands = append([]string(nil), base.ShellExec.CommandSemantics.ReadLikeCommands...)
	}

	for _, pat := range next.DangerousPathRules.WindowsRegex {
		if _, err := regexp.Compile(pat); err != nil {
			return ResourceAuditQueryConfig{}, false
		}
	}
	return next, true
}

// NormalizeResourceAuditConfig 对齐 Node normalizeResourceAuditConfig；非法 JSON 或校验失败时退回内嵌默认。
func NormalizeResourceAuditQueryConfig(raw []byte) ResourceAuditQueryConfig {
	c, ok := normalizeResourceAuditConfigBytes(raw)
	if !ok {
		return DefaultResourceAuditQueryConfig()
	}
	return c
}

// LoadResourceAuditQueryConfig 仅加载内嵌 conf/resourceaudit.toml。
func LoadResourceAuditQueryConfig() ResourceAuditQueryConfig {
	if c, ok := normalizeResourceAuditConfigBytes(conf.DefaultResourceAuditConfigTOML()); ok {
		return c
	}
	return ResourceAuditQueryConfig{}
}
